File: src/main/java/gtl/spark/java/example/C03/Cache.java
package gtl.spark.java.example.C03;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.SparkSession;

import java.util.ArrayList;
import java.util.List;

public class Cache {
    public static void main(String[] args) {
        SparkSession spark = SparkSession
                .builder()
                .master("local")
                .appName("Cache")
                .getOrCreate();
        JavaSparkContext sc = JavaSparkContext.fromSparkContext(spark.sparkContext());
        List<String> alist = new ArrayList<String>(6);
        alist.add("abc");
        alist.add("b");
        alist.add("c");
        alist.add("d");
        alist.add("e");
        alist.add("f");
        JavaRDD<String> rdd = sc.parallelize(alist);
        System.out.println(rdd);
        System.out.println("---------------");
        System.out.println(rdd.cache());
        System.out.println("---------------");
        rdd.foreach(s -> System.out.println(s));
    }
}

File: src/main/scala/gtl/spark/scala/example/C03/Cache.scala
package gtl.spark.scala.example

import org.apache.spark.{SparkConf, SparkContext}

/**
  * 提前计算的cache方法
  * cache方法的作用就是将数据内容计算并保存在计算节点的内存中。这个方法的使用是针对Spark的Lazy数据处理模式
  * 在Lazy模式中，数据在编译和未使用的时候是不进行计算的，而仅仅保存其存储地址，只有在Action方法到来的时候才正式计算。
  * 这样做的好处就是可以极大地减少存储空间，从而提高利用了，而有时必须要求数据进行计算，此时就需要使用cache方法
  **/
object Cache {
  def main(args: Array[String]): Unit = {
    val sparkConf = new SparkConf().setMaster("local").setAppName("Cache")
    val sc = new SparkContext(sparkConf)
    val arr = sc.parallelize(Array("abc", "b", "c", "d", "e", "f"))
    println(arr)
    println("------------------")
    println(arr.cache()) //这里需要理解cache、persist、checkpoint等基本概念
    println("------------------")
    arr.foreach(println) //专门用来打印为进行Action操作的数据的专用方法，额可以对数据进行提早计算
  }
}
